envoy/server/factory_context.h

```diff
@@ -272,6 +272,36 @@ class FactoryContext : public virtual ListenerAccessLogFactoryContext {
    * @return Router::Context& a reference to the router context.
    */
   virtual Router::Context& routerContext() PURE;
+
+  /**
+   * Create an FilterConfigProviderPtr for a filter config. The config providers may share
+   * the underlying subscriptions to the filter config discovery service.
+   * @param config_source supplies the extension configuration source for the filter configs.
+   * @param filter_config_name the filter config resource name.
+   * @param factory_context is the context to use for the filter config provider.
+   * @param last_filter_in_filter_chain indicates whether this filter is the last filter in the
+   * configured chain
+   * @param filter_chain_type is the filter chain type
+   * @param listener_filter_matcher is the filter matcher for TCP listener filter. nullptr for other
+   * filter types.
+   */
+  virtual void createDynamicFilterConfigProvider(
```

Comment:
Can we simplify this interface for consumption by filters? I'd expect all the contexts come from `this`, same with cluster manager. This should probably also return the provider directly rather than store the lookup function (which can affect the lifespan of the function).

Reply:
Ack. Will do the simplification.

> This should probably also return the provider directly rather than store the lookup function (which can affect the lifespan of the function).

Can you please expand this? The whole issue is we need the provider to be managed as part of FilterChainContextFactory.

Reply:
The provider manager is supposed to track subscriptions and it uses shared_ptr for the providers as a way to start/stop subscriptions. If we store persistent providers like it's done here, then the subscriptions are not ended without some explicit "deleteDynamicFilterConfigProvider", which makes for an awkward API.